Detectors/EMCAL/simulation/src/Digitizer.cxx

```diff
@@ -0,0 +1,128 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "EMCALSimulation/Digitizer.h"
+#include "EMCALBase/Digit.h"
+#include "EMCALBase/Geometry.h"
+#include "EMCALBase/GeometryBase.h"
+#include "EMCALBase/Hit.h"
+#include "MathUtils/Cartesian3D.h"
+#include "SimulationDataFormat/MCCompLabel.h"
+
+#include <TRandom.h>
+#include <TVector3.h>
+#include <climits>
+#include <forward_list>
+#include "FairLogger.h" // for LOG
+
+ClassImp(o2::EMCAL::Digitizer);
+
+// using o2::EMCAL::GeometryBase;
+using o2::EMCAL::Digit;
+using o2::EMCAL::Hit;
+
+using namespace o2::EMCAL;
+
+//_______________________________________________________________________
+void Digitizer::init() {}
+
+//_______________________________________________________________________
+void Digitizer::finish() {}
+
+//_______________________________________________________________________
+void Digitizer::process(const std::vector<Hit>* hits, std::vector<Digit>* digits)
+{
+  digits->clear();
+  mDigits.clear();
+
+  for (auto hit : *hits) {
+    Digit digit = HitToDigit(hit);
+    Int_t id = digit.GetTower();
+
+    if (id < 0 || id > mGeometry->GetNCells()) {
+      LOG(WARNING) << "tower index out of range: " << id << FairLogger::endl;
+      continue;
+    }
+
+    Bool_t flag = false;
+    for (auto rit = mDigits[id].rbegin(); rit != mDigits[id].rend(); ++rit) {
+      if (rit->CanAdd(digit)) {
+        (*rit) += digit;
+        flag = true;
+        break;
+      }
+    }
+    if (!flag)
+      mDigits[id].push_back(digit);
+  }
+
+  fillOutputContainer(digits);
+}
+
+//_______________________________________________________________________
+o2::EMCAL::Digit Digitizer::HitToDigit(Hit hit)
+{
+  TVector3 pos(hit.GetX(), hit.GetY(), hit.GetZ());
+  Int_t tower = mGeometry->GetAbsCellIdFromEtaPhi(pos.PseudoRapidity(), pos.Phi());
+  Double_t amplitude = hit.GetEnergyLoss();
+  Digit digit(tower, amplitude, mEventTime);
+  return digit;
+}
+
+//_______________________________________________________________________
+void Digitizer::setEventTime(double t)
+{
+  // assign event time, it should be in a strictly increasing order
+  // convert to ns
+  t *= mCoeffToNanoSecond;
+
+  if (t < mEventTime && mContinuous) {
+    LOG(FATAL) << "New event time (" << t << ") is < previous event time (" << mEventTime << ")" << FairLogger::endl;
+  }
+  mEventTime = t;
+}
+
+//_______________________________________________________________________
+void Digitizer::fillOutputContainer(std::vector<Digit>* digits)
```

Comment:
is it necessary to pass by pointer? It suggests that nullptr would be an allowed/expected input value - if not, it is clearer to pass by ref.

Reply:
No, indeed should be const std::vector<Digit> &. To be fixed.

Reply:
probably not const though :)